xray/src/main.rs

```diff
@@ -0,0 +1,315 @@
+mod client;
+mod key_pair;
+mod utils;
+
+use std::net::SocketAddrV4;
+
+use neptun::noise::Tunn;
+
+use tokio::{
+    net::UdpSocket,
+    sync::mpsc,
+    time::{Duration, Instant},
+};
+
+use crate::{
+    client::Client,
+    key_pair::KeyPair,
+    utils::{configure_wg, write_to_csv, Packet, RecvType, SendType, TestCmd},
+};
+
+struct CliArgs {
+    packet_count: usize,
+    csv_name: String,
+}
+
+#[tokio::main]
+async fn main() {
+    let cli_args = std::env::args().collect::<Vec<_>>();
+    let wg = cli_args
+        .get(1)
+        .map(|s| s.to_lowercase())
+        .unwrap_or_else(|| "neptun".to_owned());
+    let test_type = cli_args
+        .get(2)
+        .map(|s| s.to_lowercase())
+        .unwrap_or_else(|| "crypto".to_owned());
+    let packet_count = cli_args
+        .get(3)
+        .map(|s| {
+            s.parse::<usize>()
+                .expect("Packet count must be a positive integer")
+        })
+        .unwrap_or(10);
+    let csv_name = cli_args
+        .get(4)
+        .cloned()
+        .unwrap_or_else(|| format!("xray_metrics_{}_{}.csv", wg, test_type));
+
+    let cli_args = CliArgs {
+        packet_count,
+        csv_name,
+    };
+
+    let wg_keys = KeyPair::new();
+    let peer_keys = KeyPair::new();
+
+    let wg_name = "xraywg1";
+
+    let wg_addr: SocketAddrV4 = "100.66.0.1:41414".parse().unwrap();
+    let crypto_addr: SocketAddrV4 = "100.66.0.2:63636".parse().unwrap();
+    let crypto_sock_addr: SocketAddrV4 = "127.0.0.1:63636".parse().unwrap();
+    let pt_addr = SocketAddrV4::new(*wg_addr.ip(), 52525);
+
+    println!("Configuring wireguard with adapter type {wg}");
+    configure_wg(
+        wg.as_str(),
+        wg_name,
+        &wg_keys,
+        &peer_keys,
+        wg_addr.port(),
+        &[*wg_addr.ip()],
+    )
+    .await;
+
+    let (cmd_tx, cmd_rx) = mpsc::channel::<TestCmd>(100);
+    let (done_tx, mut done_rx) = mpsc::channel::<()>(1);
```

Comment:
I tried just awaiting the join handle before and it was exiting way too early which is why I added that channel in the first place, but I must have made some significant changes because I removed the channel now and it still works